target/linux/ramips/patches-5.4/0305-mt7621-nand-add-dma-support.patch

```diff
@@ -0,0 +1,879 @@
++		if (ret)
++			goto err;
++	}
+ 
+-	for (i = 0; i < nand->ecc.steps; i++) {
+-		if (buf)
+-			mt7621_nfc_read_data(nfc, page_data_ptr(nand, buf, i),
+-					     nand->ecc.size);
+-		else
+-			mt7621_nfc_read_data_discard(nfc, nand->ecc.size);
++err:
++	nfi_write16(nfc, NFI_CON, 0);
++	if (ret)
++		return ret;
+ 
+-		rc = mt7621_ecc_decoder_wait_done(nfc, i);
++	return nand_prog_page_end_op(nand);
++}
+ 
+-		mt7621_nfc_read_sector_fdm(nfc, i);
++static int mt7621_nfc_read_page(struct nand_chip *nand, uint8_t *buf,
++				      uint8_t *oob, int page, int hwecc)
++{
++	struct mt7621_nfc *nfc = nand_get_controller_data(nand);
++	struct mtd_info *mtd = nand_to_mtd(nand);
++	int bitflips = 0;
++	dma_addr_t addr;
++	size_t len = mtd->writesize;
++	int ret;
++	u32 reg;
+ 
+-		if (rc < 0) {
+-			bitflips = -EIO;
+-			continue;
+-		}
++	/* prepare dma */
++	addr = dma_map_single(nfc->dev, (void *)buf, len, DMA_FROM_DEVICE);
++	ret = dma_mapping_error(nfc->dev, addr);
++	if (ret) {
++		dev_err(nfc->dev, "dma mapping error\n");
++		return -EINVAL;
++	}
++	nfi_write32(nfc, NFI_STRADDR, addr);
++	init_completion(&nfc->nfi_done);
++	init_completion(&nfc->ecc_done);
+ 
+-		rc = mt7621_ecc_correct_check(nfc,
+-			buf ? page_data_ptr(nand, buf, i) : NULL,
+-			oob_fdm_ptr(nand, i), i);
+-
+-		if (rc < 0) {
+-			dev_warn(nfc->dev,
+-				 "Uncorrectable ECC error at page %d.%d\n",
+-				 page, i);
+-			bitflips = -EBADMSG;
+-			mtd->ecc_stats.failed++;
+-		} else if (bitflips >= 0) {
+-			bitflips += rc;
+-			mtd->ecc_stats.corrected += rc;
++	ret = nand_read_page_op(nand, page, 0, NULL, 0);
++	if (ret) {
++		bitflips = ret;
++		goto err;
++	}
++
++	/* init nfi/ecc */
++	reg = (CNFG_OP_CUSTOM << CNFG_OP_MODE_S) | CNFG_AUTO_FMT_EN |
++		CNFG_READ_MODE | CNFG_DMA_MODE;
++	if (hwecc)
++		reg |= CNFG_HW_ECC_EN;
++	nfi_write16(nfc, NFI_CNFG, reg);
++	if (hwecc)
++		mt7621_ecc_decoder_op(nfc, true);
++
++	/* start nfi transfer */
++	nfi_write16(nfc, NFI_CON, CON_NFI_BRD |
++		    (nand->ecc.steps << CON_NFI_SEC_S));
++	nfi_write16(nfc, NFI_STRDATA, STR_DATA);
++
++	/* wait nfi interrupt */
++	ret = wait_for_completion_timeout(&nfc->nfi_done,
++					  msecs_to_jiffies(500));
++	if (!ret) {
++		dev_warn(nfc->dev, "read ahb/dma done timeout\n");
++		bitflips = -ETIMEDOUT;
++		goto err;
++	}
++
++	if (hwecc) {
++		/* wait ecc interrupt */
++		ret = wait_for_completion_timeout(&nfc->ecc_done,
++						  msecs_to_jiffies(500));
++		if (!ret) {
++			dev_warn(nfc->dev, "decode timeout\n");
++			bitflips = -ETIMEDOUT;
++			goto err;
+ 		}
++
++		/* update ecc status */
++		bitflips = mt7621_ecc_status(nfc, mtd);
+ 	}
+ 
+-	mt7621_ecc_decoder_op(nfc, false);
++	if (oob)
++		mt7621_nfc_read_fdm(nfc, oob);
+ 
++err:
++	if (hwecc)
++		mt7621_ecc_decoder_op(nfc, false);
++	dma_unmap_single(nfc->dev, addr, len, DMA_FROM_DEVICE);
+ 	nfi_write16(nfc, NFI_CON, 0);
+ 
++	if ((bitflips >= 0) && oob) {
++		ret = mt7621_read_oob_raw(nand, oob, page, hwecc);
++		if (ret)
++			return ret;
++	}
```

Comment:
Hi!
Why is an extra read performed here? It seems to read the incorrect oob data on flash if any ecc correction happened. I can't find any reason to explicitly bypass ecc protection and return the wrong data with extra operations.

My current understanding:
1. the commit changing OOB layout added an area with no ecc protection.
2. the code here throws away ecc protection on OOB data altogether.
3. according to commit message, next commit moves bad block table out of oob area because it's now unprotected.

In order to add an extra byte into oob layout, the oob area becomes unreliable and it can't be used for bad block table. What's the benefit?

For this reason and the concerns expressed by @hackpascal , I'd also suggest dropping oob related commits.

Reply:
mtk ecc protect only total 32 bytes in oob. nand with larger oob size not protected by ecc.
for example on ubnt er-x router with nand 2048+128(32 fdm + 80 ecc + 16 remainig oob)
when use hw ecc . we only got page data(2048) and fdm data(32). but we also need ecc parity
info and remaining oob data.  so need extra read to read it out.
you can check my oob layout patch comment. the last 16 bytes 0xff oob data. is the remaining oob
data that missing.
the condition i check is not timeout(bitflips <0) and need oob info. i will do extra oob read.
the last patch comment is copy from mtk_nand.c

as @hackpascal mention. 
1. BootROM and uboot just not use first byte of fdm of subpage. but in runtime code. we can use it.
    because we will not write oob in uboot area. so BootROM can load uboot normally.
    i also check the uboot code below. it use the oob layout in write oob. seems not use in read.
    so the runtime code can boot normally. it should not cause any problem.
    https://github.com/pinney/MT7621-u-boot-mod/blob/master/drivers/nand/mtk_nand.c
    by the way. in uboot code layout. it use last oob byte to store checksum in function
    mtk_nand_write_fdm_data. but the checksum byte is not in the driver layout now.
    so the layout is not the same in uboot and kernel code.
    there also have problems in nand_oob_128. the eccbytes is 64. but the kernel driver now is 80.
    so when use old uboot with new driver . it will not boot.
2. why i change it. when use dma. hw will reorder it. but when we read/write_oob raw data. i can
    just read it out(fdm + ecc parity) every subpate. further more i can reduce code size by
    delete mt7621_nfc_write_oob_raw,mt7621_nfc_write_oob_raw and use kernel builtin 
    function nand_read_oob_syndrome,nand_write_oob_syndrome instead. this is becasue i follow
    the kernel syndrome oob layout.

Reply:
> mtk ecc protect only total 32 bytes in oob. nand with larger oob size not protected by ecc.
> for example on ubnt er-x router with nand 2048+128(32 fdm + 80 ecc + 16 remainig oob)
> when use hw ecc . we only got page data(2048) and fdm data(32). but we also need ecc parity
> info and remaining oob data. so need extra read to read it out.

For 2k+128 page, the last 16bytes in ecc mode are completely useless, and will be discarded by the hw controller.
You can't read nor write these bytes in ecc + AutoFmt mode. These bytes will never be used by mtk sdk driver and uboot.

ECC parity code in ecc mode is also useless when AutoFmt is on. In read operation, ecc parity data is sent to ecc engine for decoding, and discarded. In write operation, only main data + 4 * 8B fdm data will be sent to ecc engine for encoding, and the new ecc parity code will be written directly to NAND flash. The original ecc parity code along with the last 16b will be discarded directly.

There is completely no need to read out these data again after an ecc-enabled reading.

> you can check my oob layout patch comment. the last 16 bytes 0xff oob data. is the remaining oob
> data that missing.
> the condition i check is not timeout(bitflips <0) and need oob info. i will do extra oob read.
> the last patch comment is copy from mtk_nand.c
> 
> as @hackpascal mention.
> 
> 1. BootROM and uboot just not use first byte of fdm of subpage. but in runtime code. we can use it.
>    because we will not write oob in uboot area. so BootROM can load uboot normally.
>    i also check the uboot code below. it use the oob layout in write oob. seems not use in read.
>    so the runtime code can boot normally. it should not cause any problem.
>    https://github.com/pinney/MT7621-u-boot-mod/blob/master/drivers/nand/mtk_nand.c
>    by the way. in uboot code layout. it use last oob byte to store checksum in function
>    mtk_nand_write_fdm_data. but the checksum byte is not in the driver layout now.
>    so the layout is not the same in uboot and kernel code.

Checksum code is only calculated by SW and not used. It's obsolete.

>    there also have problems in nand_oob_128. the eccbytes is 64. but the kernel driver now is 80.
>    so when use old uboot with new driver . it will not boot.

Because 80 is the right value.
For 2k+128 page:
ECC size = 512
ECC steps = 2k / 512 = 4
Spare size for each step = 128 / 4 = 32
However, the MT7621 NFI module on supports a maximum spare size of 28 bytes
The fdm size of each step is 8 bytes
So the size of ecc parity code for each step = 28 - 8 = 20 bytes
For a subpage with size = 512, using BCH, one bit correction will cost 13 bits.
The MT7621 ECC engine supports a maximum BCH correction strength with 12 bits.
12 bits correction capability will cost total 12 * 13 = 156 bits = 19 bytes + 4 bits = rounded up to 20 bytes
So total ecc bytes = 4 * 20 = 80 bytes

> 2. why i change it. when use dma. hw will reorder it. but when we read/write_oob raw data. i can
>    just read it out(fdm + ecc parity) every subpate. further more i can reduce code size by
>    delete mt7621_nfc_write_oob_raw,mt7621_nfc_write_oob_raw and use kernel builtin
>    function nand_read_oob_syndrome,nand_write_oob_syndrome instead. this is becasue i follow
>    the kernel syndrome oob layout.

What you're saying only applies to raw mode. You made this driver using two different data organizing flows for ecc mode and raw mode, by introducing more modifications. Since the ecc mode can't follow kernel's syndrome flow, it's better to just change the DMA operation of raw mode, and keep the original code flow which is also aligned with mtk_nand.c

Reply:
> we only got page data(2048) and fdm data(32). but we also need ecc parity
> info and remaining oob data. so need extra read to read it out.

nobody need that ecc parity info. It can only be understood by the controller to perform a correction and it's meaningless to anybody else.

> you can check my oob layout patch comment. the last 16 bytes 0xff oob data. is the remaining oob
> data that missing.

and these are fixed 0xff bytes if you don't write extra code in writing routine to use them, so it's also meaningless.

> the condition i check is not timeout(bitflips <0) and need oob info. i will do extra oob read.
> the last patch comment is copy from mtk_nand.c

Oops. I misread that one.
but the code is still pointless: hardware already did the reading of oob and ecc correction. You throw all these work away, perform at most 4 more reading operations and reorganize these data manually in software to get some meaningless data.

> 2. why i change it. when use dma. hw will reorder it. but when we read/write_oob raw data. i can
>    just read it out(fdm + ecc parity) every subpate. further more i can reduce code size by
>    delete mt7621_nfc_write_oob_raw,mt7621_nfc_write_oob_raw and use kernel builtin
>    function nand_read_oob_syndrome,nand_write_oob_syndrome instead. this is becasue i follow
>    the kernel syndrome oob layout.

kernel can't take interleaved data/oob so a data reorder is required anyway. you are just replacing one reorder routine with another.

what's the benefit of these oob changes for openwrt?
we get a more complex reading routine with extra operations and the we lost ecc protection of usable oob areas in order to read some useless oob data. It isn't useful for openwrt as openwrt neither perform nand operation with ecc disabled, nor can it make use of any extra data read with these oob changes.

Reply:
the benefit is follow Linux kernel api requirement.
it need all oob data. i will provide it to kernel. if kernel don't need oob. i don't do extra read.
so for normal read operation. the performance is the same. because normal read don't need
oob data.

what cause a more complex reading routing. it's mtk's hardware design. make me do these
work to fit kernel requirement. oob data shouldn't be ecc protected. it cause not fit Linux framework.
remember the jffs2 workaround i remove. you can find these workaround in mtk router sdk.
so the driver have bug. because it not support jffs2.

we don't lost ecc protection. because it only protect 32 bytes fdm data in oob. and not protect others.
i just do kernel ask me to do. read remaining oob data.

about the layout. mtk follow the hw ecc layout (look like HW_ECC layout). i follow the raw layout and it
is the syndrome layout should look like. just the choices problem. we all need to reorder oob data.

so the every first byte of fdm of subpage will cause any problem when i use it?

Reply:
> what cause a more complex reading routing. it's mtk's hardware design. make me do these
> work to fit kernel requirement.

The original approach is already used in upstream mtk_nand driver for mtk arm socs, which means it's already approved by upstream folks and the "requirement" you are trying to fit into just doesn't exist. There's no need to introduce a slower and more complex oob reading routine when the original one just do it's work fine.

> oob data shouldn't be ecc protected. it cause not fit Linux framework.

There's no such a requirement either. There are many controllers providing ecc protected usable oob area and we don't need to bypass it.

> remember the jffs2 workaround i remove. you can find these workaround in mtk router sdk.
> so the driver have bug. because it not support jffs2.

General approach for downstream patches in OpenWrt:
1. If the patch can go upstream, send it upstream and backport it after it's approved there.
2. If it can't go there, we only take it if it's necessary for OpenWrt.

The nand driver we currently use is rejected by upstream because it's similar to existing mtk_nand.c and upstream folks don't want this kind of code duplication.
We still take the driver because the old nand driver we use is nasty and it needs several hacks in kernel nand framework. This driver replacement relieves a huge maintenance burden of porting those hacks, which benefits future kernel updating for OpenWrt.

I rejected the jffs2 workaround because OpenWrt avoids using jffs2 on nand anyway. We don't need this workaround so I don't want to introduce it as one more downstream patch.
All these oob changes with added code complexity are made to fit some nonexistent requirement, and they don't benefit OpenWrt either. so I reject it here as well.

> 
> we don't lost ecc protection. because it only protect 32 bytes fdm data in oob. and not protect others.

That fdm data is the only usable area and you trade the protection for it to read other useless data. It's pointless.

> i just do kernel ask me to do. read remaining oob data.

There's no way of using those unprotected oob data anyway, and a driver designed to use oob area will not touch these data unless they are claimed as free area and someone adds an extra writing routine for it. (To be clear: I'm not suggesting anyone to write this routine. we don't need it.)

> 
> about the layout. mtk follow the hw ecc layout (look like HW_ECC layout). i follow the raw layout and it
> is the syndrome layout should look like. just the choices problem. we all need to reorder oob data.

The raw layout is subpages with interleaved data/oob and there's currently no way for kernel to have a matching oob layout. As both layout don't match the actual layout anyway, this switch achieved nothing.

> so the every first byte of fdm of subpage will cause any problem when i use it?

This introduced a potential breakage as already said by @hackpascal (I don't think checking one u-boot driver and assume that nobody writes oob in u-boot area are enough to say it's safe to do so.) and it doesn't bring any value for openwrt. I'd prefer not having this change.

Reply:
the upstream mtk_nand.c only protect 1 byte fdm data per subpage. but mt7621 protect 8 bytes fdm data per subpage. on arm version mark fdm data is not free oob area. so program can't use it. the remaining oob
is not protect. is also cause why arm version don't have jffs2 problem. because jffs2 writes to not protect
oob data area. if ecc should be protect why mtk change fdm size from 8 to 1. because it cause problem.

i mention jffs2 because this is the way how kernel use oob. i use it to verify my patch. because i don't know
there may be other modules use the same way to access the driver. i wan the driver bug free.

could you provide which nand driver protect oob data and also let this area writable. as i know mtk_nand.c is not on the list.

when i use nanddump oob data with ecc. it only show fdm data. but not show ecc parity and remaining oob
data. when i use nanddump oob data without ecc. it will show fdm data and ecc parity. but not show remaining oob data. these are problems. i fix it by add a function to read it. you can check kernel default
function nand_read_oob_syndrome i copy from. kernel use the same way to read oob data. it a simple function not complexity. it read all oob data out. i just do what the function should do.

we can use remaining oob data by nandwrite. and bbt(bad block table) function also want to use it.
but the driver can't provide all the oob data for bbt. so we have the last patch. told bbt not use oob.
so these oob are not useless. just because you haven't use it. i use nanddump/nandwrite to read/write oob data. to verify my patch is correct. maybe for you not use this function. so you think it is useless. but i implement it to help me to debug driver.

i didn't change any fdm function. i just do more oob read if user require it.

Reply:
> if ecc should be protect why mtk change fdm size from 8 to 1. because it cause problem.

That's just a per-SoC option to match bootrom choice. It can only be configured the same way as bootrom or bootrom won't be able to load u-boot.
 
> could you provide which nand driver protect oob data and also let this area writable. as i know mtk_nand.c is not on the list.

Gigadevice and Winbond SPI NAND chips both provides ECC-protected oob areas. A SPI NAND is effectively a raw nand with a spi-attached nand controller.

> we can use remaining oob data by nandwrite.

That area isn't part of available oob space so kernel treats it as reserved area and won't write data to it when drivers want to write their own data to oob. It doesn't matter how you could write to it by using nandwrite in raw mode, effectively skips kernel oob checking.

> and bbt(bad block table) function also want to use it.

but your implementation doesn't allow it to be used.

> but the driver can't provide all the oob data for bbt. so we have the last patch. told bbt not use oob.
> so these oob are not useless. just because you haven't use it. i use nanddump/nandwrite to read/write oob data. to verify my patch is correct. maybe for you not use this function. so you think it is useless. but i implement it to help me to debug driver.

You are submitting a patch for openwrt to review/maintain, so you need to explain how openwrt could benefit from these commits, not how it helps you during debugging. Why should openwrt take a dozen lines of code when it can't benefit from the added code at all?

Reply:
BTW It's still arguable whether the jffs2 fix is needed but that jffs2 problem has no relation with oob layout change we are currently discussing.
NAK on oob layout change because upstream already approved the same approach to current oob layout in mtk_nand.c and the added ability of reading ecc parity data and unusable oob area are pointless.

Reply:
i also use winbond spi nand(ex W25N01GWxxIG/IT) before. as i know it provide two mode. internal ecc mode. use it like spi nor flash. let spi internal ecc to correct data. when use this mode. it has two ecc area. one protect main page data. another protect some oob data and previous ecc area. another mode is disable internal ecc. use as normal nand flash. let external controller to write ecc data.
so what we talk is raw nand flash driver (in raw directory). not spi nand flash.

because i add remaining oob area to ooblayout free area. so i can use nanddump/nandwrite to access remaining oob data.

so as i say we have the last patch to fix bbt problem. just like mtk_nand.c. by the way no one well care it
because mtk use private bbt method. not linux version. so you don't need this patch too? because no one will use it.

it's not only for debug. kernel provide this api. i complete implement it. so nanddump/nandwrite to access as much more available free oob area as possible. the old driver only support part of oob area in large oob flash.

if NAK i just give up these patches.